evm/deploy-gateway-v5.0.x.js

```diff
@@ -341,10 +366,63 @@ async function upgrade(config, options) {
         if (anwser !== 'y') return;
     }
 
-    const tx = await gateway.upgrade(contractConfig.implementation, implementationCodehash, setupParams, gasOptions);
-    printInfo('Upgrade transaction', tx.hash);
+    // Offline signing
+    let gasLimit, gasPrice;
 
-    await tx.wait(chain.confirmations);
+    try {
+        gasPrice = parseUnits((await provider.getGasPrice()).toString(), 'gwei');
+        const block = await provider.getBlock('latest');
+        gasLimit = block.gasLimit.toNumber() / 1000;
+
+        printInfo('Gas Price:', gasPrice.toString());
+        printInfo('Gas Limit:', gasLimit.toString());
+    } catch (error) {
+        printError('Gas price and limit could not be fetched from provider');
+        printObj(error);
+    }
+
+    let nonce = await getNonceFromProvider(provider, signerAddress);
+    let signersData, transactions;
+
+    if (isOffline) {
+        directoryPath = directoryPath || './txs';
+        fileName = fileName || env.toLowerCase() + '-' + chain.name.toLowerCase() + '-' + 'signedUpgradeTransactions';
+        printInfo(`Storing signed Txs offline in file ${fileName}`);
+        nonce = await getLatestNonceAndUpdateData(directoryPath, fileName, wallet);
+        signersData = await getAllSignersData(directoryPath, fileName);
+        transactions = await getTransactions(directoryPath, fileName, signerAddress);
+        const { baseTx, signedTx } = await ledgerSign(
```

Comment:
this is complicated to work with.

Ledger sign util should just take a tx and a chain object, populate any fields like chainId, gasPrice, nonce etc if missing to keep ux simpler